### PR TITLE
GH-38998: [Java] Build memory-core and memory-unsafe as JPMS modules

### LINKED ISSUE #38998: [Java] Build memory-core and memory-unsafe as JPMS modules
### Describe the enhancement requested
This depends on #39134 .

Build memory-core and memory-unsafe as JPMS modules.

- Refactor the memory module's structures to faciliate this.

These modules use Unsafe and reflection and require adding --add-opens command line flags since they are now named modules.

Note that memory-netty is omitted because it requires a bigger change to the user's command-line.

### Component(s)

Java

### PR DESCRIPTION
**Pull request by jduo**
Depends on #39134 

### Rationale for this change
Part of modularizing Arrow. Allows these components to be used by tools that require JPMS modules such as jlink.

### What changes are included in this PR?
- Refactor memory modules so that all netty code is in netty-memory.
- All code being injected into Netty packages and have dependencies on Netty's package-private code go into a new module memory-netty-buffer-patch.
- The surefire command line has been changed to allow reflection on unsafe to be used by arrow-memory-core
- Add module-info files for arrow-memory-core and arrow-memory-unsafe

### Are these changes tested?
Yes, existing tests work when run with modules.

### Are there any user-facing changes?
Yes, users now need to put --add-opens=java.base/java.nio=org.apache.arrow.memory.core,ALL-UNNAMED on their java command line instead of --add-opens=java.base/java.nio=ALL-UNNAMED since memory-core is now a named module instead of unnamed.

 **This PR includes breaking changes to public APIs.** 
Yes, users now need to put --add-opens=java.base/java.nio=org.apache.arrow.memory.core,ALL-UNNAMED on their java command line instead of --add-opens=java.base/java.nio=ALL-UNNAMED since memory-core is now a named module instead of unnamed.
* Closes: #38998

### CODE REVIEW

**Comment by jduo** [comment]
It seems these changes are not usable without modularizing memory-core. arrow-vector makes use of Preconditions and BufferAllocator heavily. Since these are declared in an unnamed module (memory-core), the user has to add --open-reads=org.apache.arrow.vector=ALL-UNNAMED. This doesn't fail when running arrow-vector tests because surefire adds this automatically for this module but fails any other test that depends on arrow-vector.

The Preconditions class could be moved to a new utils module, but that doesn't deal with BufferAllocator which is very much part of memory-core out of necessity.

File: java/maven/module-info-compiler-maven-plugin/pom.xml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+              <goal>helpmojo</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+
+
+</project>
```
Comment (vibhatha) [inline java/maven/module-info-compiler-maven-plugin/pom.xml:130]: new line?

File: java/maven/module-info-compiler-maven-plugin/src/main/java/org/apache/arrow/maven/plugins/BaseModuleInfoCompilerPlugin.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      } catch (IOException ex) {
+        throw new MojoExecutionException("Error compiling module-info.java", ex);
+      }
+    } else {
+      getLog().info("No module-info.java file found. module-info.class file was not generated.");
+    }
+  }
+
+  /**
+   * Finds the first module-info.java file in the set of source directories.
```
Comment (vibhatha) [inline java/maven/module-info-compiler-maven-plugin/src/main/java/org/apache/arrow/maven/plugins/BaseModuleInfoCompilerPlugin.java:77]: @jduo Just for the curiosity, why are we just checking the first `module-info.java`?

File: java/memory/memory-core/src/main/java/org/apache/arrow/memory/util/MemoryUtil.java [hunk 3]
```diff
@@ -142,7 +142,7 @@ public Object run() {
       // the static fields above get initialized
       final RuntimeException failure = new RuntimeException(
           "Failed to initialize MemoryUtil. You must start Java with " +
-              "`--add-opens=java.base/java.nio=ALL-UNNAMED` " +
+              "`--add-opens=java.base/java.nio=org.apache.arrow.memory.core,ALL-UNNAMED` " +
```
Comment (vibhatha) [inline java/memory/memory-core/src/main/java/org/apache/arrow/memory/util/MemoryUtil.java:145]: would this require a documentation update?

File: java/memory/memory-core/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java [hunk 4]
```diff
@@ -448,73 +443,73 @@ public ArrowBuf empty() {
   @Test
   public void testRootAllocator_listeners() throws Exception {
     CountingAllocationListener l1 = new CountingAllocationListener();
-    assertEquals(0, l1.getNumPreCalls());
-    assertEquals(0, l1.getNumCalls());
-    assertEquals(0, l1.getNumReleaseCalls());
-    assertEquals(0, l1.getNumChildren());
-    assertEquals(0, l1.getTotalMem());
+    Assert.assertEquals(0, l1.getNumPreCalls());
```
Comment (vibhatha) [inline java/memory/memory-core/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java:446]: nit: should we make `l1` -> `listner1`. I know it is not related to this PR `11` vs `l1`, just for the readability. And may be the following names too?

File: java/memory/memory-core/src/test/java/org/apache/arrow/memory/TestForeignAllocation.java [hunk 5]
```diff
@@ -23,6 +23,13 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import org.apache.arrow.memory.AllocationListener;
+import org.apache.arrow.memory.AllocationOutcome;
+import org.apache.arrow.memory.ArrowBuf;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.ForeignAllocation;
+import org.apache.arrow.memory.OutOfMemoryException;
+import org.apache.arrow.memory.RootAllocator;
```
Comment (vibhatha) [inline java/memory/memory-core/src/test/java/org/apache/arrow/memory/TestForeignAllocation.java:32]: seems like these are not used.

**Review by vibhatha — COMMENTED** [review]
@jduo I have added a few comments.

File: java/maven/module-info-compiler-maven-plugin/src/main/java/org/apache/arrow/maven/plugins/BaseModuleInfoCompilerPlugin.java (hunk 2 above)
Comment (jduo) [reply java/maven/module-info-compiler-maven-plugin/src/main/java/org/apache/arrow/maven/plugins/BaseModuleInfoCompilerPlugin.java:77]: A JAR can only have one module-info file. We are arbitrarily taking the first. I haven't checked if the maven-compiler-plugin has the same behavior or if it errors out when there are multiple.

File: java/pom.xml [hunk 6]
```diff
@@ -870,7 +897,7 @@
             <groupId>org.apache.maven.plugins</groupId>
             <artifactId>maven-surefire-plugin</artifactId>
             <configuration>
-              <argLine>--add-opens=java.base/java.nio=ALL-UNNAMED</argLine>
+              <argLine>--add-opens=java.base/java.nio=org.apache.arrow.memory.core,ALL-UNNAMED</argLine>
```
Comment (lidavidm) [inline java/pom.xml:900]: Just to be sure, do we still need ALL-UNNAMED?

File: java/memory/memory-unsafe/src/test/java/org/apache/arrow/memory/unsafe/TestUnsafeAllocationManager.java [hunk 7]
```diff
@@ -15,20 +15,25 @@
...
 import org.junit.Test;
 
 /**
  * Test cases for {@link UnsafeAllocationManager}.
  */
 public class TestUnsafeAllocationManager {
 
-  private BaseAllocator createUnsafeAllocator() {
-    return new RootAllocator(BaseAllocator.configBuilder().allocationManagerFactory(UnsafeAllocationManager.FACTORY)
+  private RootAllocator createUnsafeAllocator() {
```
Comment (lidavidm) [inline java/memory/memory-unsafe/src/test/java/org/apache/arrow/memory/unsafe/TestUnsafeAllocationManager.java:35]: Ideally we never refer to RootAllocator (or BaseAllocator) as a type, only BufferAllocator

File: java/memory/memory-core/pom.xml [hunk 8]
```diff
@@ -54,6 +53,30 @@
   </build>
 
   <profiles>
+    <profile>
+      <id>error-prone-jdk11+</id>
```
Comment (lidavidm) [inline java/memory/memory-core/pom.xml:57]: how is this profile related to ErrorProne?

File: java/pom.xml (hunk 6 above)
Comment (jduo) [reply java/pom.xml:900]: Yes, it's still needed. Netty utilizes it for example.

File: java/memory/memory-core/pom.xml (hunk 8 above)
Comment (jduo) [reply java/memory/memory-core/pom.xml:57]: This profile is really the "default" build profile for JDK9+. It doesn't relate to error-prone specifically. Would it be better to have a separate profile with the same activation?

File: java/memory/memory-core/src/main/java/org/apache/arrow/memory/util/MemoryUtil.java (hunk 3 above)
Comment (vibhatha) [reply java/memory/memory-core/src/main/java/org/apache/arrow/memory/util/MemoryUtil.java:145]: should we include it to this PR or file a separate issue for that?

File: java/maven/module-info-compiler-maven-plugin/src/main/java/org/apache/arrow/maven/plugins/BaseModuleInfoCompilerPlugin.java (hunk 2 above)
Comment (vibhatha) [reply java/maven/module-info-compiler-maven-plugin/src/main/java/org/apache/arrow/maven/plugins/BaseModuleInfoCompilerPlugin.java:77]: Noted, thanks.

File: java/memory/memory-core/pom.xml (hunk 8 above)
Comment (lidavidm) [reply java/memory/memory-core/pom.xml:57]: Ah, then just rename it so it's clear - the "error-prone" profiles we have are meant for running the ErrorProne static analyzer

File: java/memory/memory-core/src/main/java/org/apache/arrow/memory/util/MemoryUtil.java (hunk 3 above)
Comment (jduo) [reply java/memory/memory-core/src/main/java/org/apache/arrow/memory/util/MemoryUtil.java:145]: I think it's best to do a separate PR for doc updates around this since there are a few of these across multiple PRs.

File: java/maven/module-info-compiler-maven-plugin/pom.xml (hunk 1 above)
Comment (jduo) [reply java/maven/module-info-compiler-maven-plugin/pom.xml]: Fixed now.

File: java/memory/memory-core/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java (hunk 4 above)
Comment (jduo) [reply java/memory/memory-core/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java]: Done.

File: java/memory/memory-unsafe/src/test/java/org/apache/arrow/memory/unsafe/TestUnsafeAllocationManager.java (hunk 7 above)
Comment (jduo) [reply java/memory/memory-unsafe/src/test/java/org/apache/arrow/memory/unsafe/TestUnsafeAllocationManager.java]: Changed to return BufferAllocator instead.

File: java/memory/memory-core/pom.xml (hunk 8 above)
Comment (jduo) [reply java/memory/memory-core/pom.xml]: Changed the profile to be jdk11+.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
If we're going to include this with Arrow 15, we should get on the doc changes immediately.

**Comment by jduo** [comment]
> If we're going to include this with Arrow 15, we should get on the doc changes immediately.

OK, I have added the doc change to this PR instead.

**Review by lidavidm — APPROVED** [review]

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java* *verification*

**Comment by lidavidm** [comment]
Sorry, I didn't get around to this and the code freeze is today...this might have to slip to arrow 16 since I realized we never checked the other Java builds

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java* *verify*

**Comment by lidavidm** [comment]
It appears Javadocs builds fail:

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.3:jar (attach-javadocs) on project arrow-memory-core: MavenReportException: Error while generating Javadoc: 
Error:  Exit code: 1 - /Users/runner/work/crossbow/crossbow/arrow/java/memory/memory-core/src/main/java/module-info.java:18: error: class, interface, or enum expected
Error:  module org.apache.arrow.memory.core {
Error:  ^
```

**Comment by jduo** [comment]
> Sorry, I didn't get around to this and the code freeze is today...this might have to slip to arrow 16 since I realized we never checked the other Java builds

No problem. It is probably best this gets completed in the same release as #39001 too so this gives a bit more time to examine that as well.

**Comment by jduo** [comment]
@github-actions crossbow submit *java*

**Comment by jduo** [comment]
@github-actions crossbow submit *java* *verify*

**Comment by jduo** [comment]
@github-actions crossbow submit *java* *verify*

**Comment by jduo** [comment]
@github-actions crossbow submit *java* *verify*

**Comment by jduo** [comment]
@github-actions crossbow submit *java* *verify*

**Comment by jduo** [comment]
@lidavidm , the javadoc generation failure is now fixed. I updated the maven-javadoc-plugin to a newer version that properly let you exclude files and excluded the module-info.java files.

**Comment by jduo** [comment]
> It appears Javadocs builds fail:
> 
> ```
> Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.3:jar (attach-javadocs) on project arrow-memory-core: MavenReportException: Error while generating Javadoc: 
> Error:  Exit code: 1 - /Users/runner/work/crossbow/crossbow/arrow/java/memory/memory-core/src/main/java/module-info.java:18: error: class, interface, or enum expected
> Error:  module org.apache.arrow.memory.core {
> Error:  ^
> ```

This issue is now fixed by updating maven-javadoc-plugin to a version that allows source exclusions properly and excluding module-info.java files.

File: java/pom.xml [hunk 9]
```diff
@@ -776,6 +787,11 @@
             </reports>
           </reportSet>
         </reportSets>
+        <configuration>
+          <sourceFileExcludes>
+            <sourceFileExclude>**/module-info.java</sourceFileExclude>
+          </sourceFileExcludes>
+        </configuration>
```
Comment (lidavidm) [inline java/pom.xml:794]: We may want to document module-info.java in Java 9+ right?

**Review by lidavidm — APPROVED** [review]

File: java/pom.xml (hunk 9 above)
Comment (jduo) [reply java/pom.xml:801]: Yes, I think we should wait until JDK8 support is dropped to do this though. We run into similar issues as we ran into while compiling (need to set release to 9, but that breaks how we utilize Unsafe).

File: java/memory/memory-core/src/test/java/org/apache/arrow/memory/CountingAllocationListener.java [hunk 10]
```diff
@@ -58,6 +64,10 @@ public void onAllocation(long size) {
   @Override
   public boolean onFailedAllocation(long size, AllocationOutcome outcome) {
     if (expandOnFail) {
+      if (expandAlloc == null) {
```
Comment (lidavidm) [inline java/memory/memory-core/src/test/java/org/apache/arrow/memory/CountingAllocationListener.java:67]: Could possibly use Preconditions.checkState for this (since it's been annotated such that CheckerFramework should understand it)

**Review by lidavidm — APPROVED** [review]
